R/tmb-methods.R

```diff
@@ -40,43 +40,146 @@ fitted.mmrm_tmb <- function(object, ...) {
+  return(res)
+}
+
+#' Get Prediction
+#' @description Get prediction with given data, theta, beta, beta_vcov
+#'
+#' @param tmb_data (`mmrm_tmb_data`)\cr object.
+#' @param theta (`numeric`)\cr theta value.
+#' @param beta (`numeric`)\cr beta value.
+#' @param beta_vcov (`matrix`)\cr beta_vcov matrix.
+#'
+#' @keywords internal
+h_get_prediction <- function(tmb_data, theta, beta, beta_vcov) {
+  assert_class(tmb_data, "mmrm_tmb_data")
+  assert_numeric(theta)
+  n_beta <- ncol(tmb_data$x_matrix)
+  assert_numeric(beta, finite = TRUE, any.missing = FALSE, len = n_beta)
+  assert_matrix(beta_vcov, mode = "numeric", any.missing = FALSE, nrows = n_beta, ncols = n_beta)
+  .Call(`_mmrm_predict`, PACKAGE = "mmrm", tmb_data, theta, beta, beta_vcov)
+}
+
+#' Get Prediction Variance
+#' @description Get prediction variance with given fit, tmb_data with sampling method
+#'
+#' @param object (`mmrm_tmb`)\cr the fitted MMRM.
+#' @param n_sim (`integer`)\cr number of replication of sampling.
+#' @param tmb_data (`mmrm_tmb_data`)\cr object.
+#'
+#' @keywords internal
+h_get_prediction_variance <- function(object, n_sim, tmb_data) {
+  assert_class(object, "mmrm_tmb")
+  assert_class(tmb_data, "mmrm_tmb_data")
+  assert_int(n_sim)
+  theta_chol <- chol(object$theta_vcov)
+  n_theta <- length(object$theta_est)
+  res <- replicate(n_sim, {
+    z <- rnorm(n = n_theta)
+    theta_sample <- object$theta_est + theta_chol %*% z
+    cond_beta_results <- object$tmb_object$report(theta_sample)
+    beta_mean <- cond_beta_results$beta
+    beta_cov <- cond_beta_results$beta_vcov
+    h_get_prediction(tmb_data, theta_sample, beta_mean, beta_cov)
+  })
+  mean_of_var <- rowMeans(res[, 1, ])
+  var_of_mean <- apply(res[, 3, ], 1, var)
+  mean_of_var + var_of_mean
+}
+
+
 #' @describeIn mmrm_tmb_methods obtains the model frame.
-#' @param exclude (`character`)\cr names of variable to exclude.
+#' @param include (`character`)\cr names of variable to include.
 #' @param full (`flag`) indicator whether to return full model frame (deprecated).
+#' @param na.action (`string`) na action.
 #' @importFrom stats model.frame
 #' @exportS3Method
 #'
 #' @details
-#' `exclude` argument controls the variables the returned model frame will exclude.
-#' Possible options are "subject_var", "visit_var" and "group_var", representing the
-#' subject variable, visit variable or group variable.
+#' `include` argument controls the variables the returned model frame will include.
+#' Possible options are "response_var", "subject_var", "visit_var" and "group_var", representing the
+#' response variable, subject variable, visit variable or group variable.
 #'
 #' @examples
 #' # Model frame:
 #' model.frame(object)
-#' model.frame(object, exclude = "subject_var")
-model.frame.mmrm_tmb <- function(formula, exclude = "subject_var", full, ...) {
+#' model.frame(object, include = "subject_var")
+model.frame.mmrm_tmb <- function(formula, include = NULL, full, na.action = "na.omit", ...) { # nolint
```

Comment:
here it will be important to have robust testing coverage including edge cases, since now it is quite complex